R/risk_mod.R

```diff
@@ -276,62 +283,88 @@ risk_mod <- function(X, y, gamma = NULL, beta = NULL, weights = NULL,
+      gamma <- upd$gamma
+      beta <- upd$beta
+    }
 
-  # Initial beta is null then round LR coefficients using median
-  if (is.null(beta)){
-    # Initial model
-    df <- data.frame(X, y)
-    init_mod <- stats::glm(y~.-1, family = "binomial", weights = weights, data = df)
+    # Initial beta is null then round LR coefficients using median
+    if (is.null(beta)){
+      # Initial model
+      df <- data.frame(X, y)
+      init_mod <- stats::glm(y~.-1, family = "binomial", weights = weights, data = df)
 
-    # Replace NA's with 0's
-    coef_vals <- unname(stats::coef(init_mod))
-    coef_vals[is.na(coef_vals)] <- 0
+      # Replace NA's with 0's
+      coef_vals <- unname(stats::coef(init_mod))
+      coef_vals[is.na(coef_vals)] <- 0
 
-    # Round so betas within range
-    gamma <- max(abs(coef_vals[-1]))/min(abs(a + 0.5), abs(b + 0.5))
-    beta <- coef_vals/gamma
-    beta <- randomized_rounding(beta)
-  }
+      # Round so betas within range
+      gamma <- max(abs(coef_vals[-1]))/min(abs(a + 0.5), abs(b + 0.5))
+      beta <- coef_vals/gamma
+      beta <- randomized_rounding(beta)
+    }
 
+    # Check no numeric issues
+    if (is.nan(gamma) | sum(is.nan(beta)) > 0){
+      stop("Initial gamma or beta is NaN - check starting value for beta")
+    }
+    if (is.na(gamma) | sum(is.na(beta)) > 0){
+      stop("Initial gamma or beta is NA - check starting value for beta")
+    }
+    if (length(beta) != ncol(X)) stop("beta and X non-compatible")
+    if (length(y) != nrow(X)) stop("y and X non-compatible")
 
-  # Check no numeric issues
-  if (is.nan(gamma) | sum(is.nan(beta)) > 0){
-    stop("Initial gamma or beta is NaN - check starting value for beta")
-  }
-  if (is.na(gamma) | sum(is.na(beta)) > 0){
-    stop("Initial gamma or beta is NA - check starting value for beta")
+    # Run coordinate descent from initial solution
+    res <- risk_coord_desc(X, y, gamma, beta, weights, lambda0, a, b, max_iters,
+                           tol, shuffle)
+
+    gamma <- res$gamma
+    beta <- res$beta
+
+    return(list(gamma=gamma, beta=beta, X=X, y=y,
+                weights=weights, lambda0=lambda0))
   }
-  if (length(beta) != ncol(X)) stop("beta and X non-compatible")
-  if (length(y) != nrow(X)) stop("y and X non-compatible")
 
-  # Run coordinate descent from initial solution
-  res <- risk_coord_desc(X, y, gamma, beta, weights, lambda0, a, b, max_iters,
-                         tol, shuffle)
-  gamma <- res$gamma
-  beta <- res$beta
+  # Track minimum objective function and best model parameters
+  min_obj_fn <- Inf
+  best_gamma <- NULL
+  best_beta <- NULL
+
+  # Run n_train_runs to find the best model
+  for (i in 1:n_train_runs) {
+    curr_mod <- run_risk_mod(X, y, gamma, beta, weights, lambda0, a, b,
+                             max_iters, tol, shuffle)
+    curr_obj_fn <- obj_fcn(curr_mod$X, curr_mod$y, curr_mod$gamma,
+                           curr_mod$beta, curr_mod$weights, curr_mod$lambda0)
```

Comment:
**question**: do we need `curr_mod` for `X`, `y`, `weights` and `lambda0`? Can't we just use the same ones from the original function call since they don't change in the `run_risk_mod()` initialization?